server/src/c0Completions.ts

```diff
@@ -0,0 +1,118 @@
+import * as vscode from "vscode-languageserver";
+import { fromVscodePosition, toVscodePosition } from "./ast";
+import * as nearley from "nearley";
+import * as exp from "./expression-rules";
+import * as parsed from "./parse/parsedsyntax";
+
+export const enum CompletionContextKind {
+  StructAccess,
+  FunctionCall
+}
+
+export type CompletionResult =
+  | StructAccess
+  | FunctionCall
+  | null;
+
+export interface StructAccess {
+  tag: CompletionContextKind.StructAccess;
+  expr: parsed.Expression;
+}
+
+export interface FunctionCall {
+  tag: CompletionContextKind.FunctionCall;
+  name: string;
+  argumentNumber: number;
+}
+
+function scanExpression(source: string, index: number) {
+  let parenStack = 0;
+
+  let pos;
+  for (pos = index; pos >= 0; pos--) {
+    const c = source[pos];
+
+    if (c === ")") parenStack++;
+    if (c === "(") {
+      if (parenStack === 0) break; else parenStack--;
+    }
+    if (c === "=") break;
+    if (c === ";") break;
+    if (c === ",") break;
+
+    if (source.startsWith("return", pos - 6)) break;
+  }
+  return pos < 0 ? "" : source.slice(pos + 1, index).trim();
+}
+
+function scanFunctionName(source: string, index: number) {
+  let pos;
+  for (pos = index; pos >= 0; pos--) {
+    if (!/[A-Za-z0-9_]/.test(source[pos])) break;
+  }
+
+  // Add one, `index` is the last character of the function name 
+  return pos < 0 ? "" : source.slice(pos + 1, index + 1).trim();
+}
+
+export function getCompletionContext(source: string, index: number): CompletionResult {
+  const parser = new nearley.Parser(nearley.Grammar.fromCompiled(exp.default));
+  // Ignore errors 
+  (<any>parser).reportError = () => "";
+  
+  // Meaningful character to use are either:
+  // - left paren (function call or casted expression)
+  // - comma (function argument, we can skip to the left paren and then use that)
+  // - struct dereference (-> or .)
+
+  let pos = index;
+  // Skip whitespace
+  while (source[pos] === " ") pos--;
+
+  // Struct access must be at the cursor (barring whitespace)
+  if (source.startsWith("->", pos - 2)) {
```

Comment:
Correct me if I'm wrong but I think you're missing the case for '.'

Reply:
Lmao totally forgot that was a thing. It won't work for arrays of structs but its a start